Src/ST-LIB_LOW/StateMachine/StateMachine.cpp

```diff
@@ -105,6 +109,13 @@ StateMachine::StateMachine(uint8_t initial_state) :
 	initial_state(initial_state), current_state(initial_state) {
 	add_state(initial_state);
 	enter_state(initial_state);
+
+	#ifdef SIM_ON
+		SharedMemory::start_emulated_state_machine();
+		uint8_t counter=SharedMemory::state_machine_sm[0]+1;
```

Comment:
store this counter as part of the state machine, later you are searching for the old state, but multiple state machines might have the same ID for a state, which won't work with your logic

Reply:
What do you mean by storing the name as part of the state machine, creating a public instance variable at StateMachine.hpp that points to that value???

Reply:
to create a new private member of the StateMachine class named something like `shm_id` where you store the counter and then you use that variable to access the SHM

Reply:
no, I guess more like this:
```c++
uint8_t state_machine_id_in_shm;
static uint8_t state_machines_count{0};
```
In constructor:
```c++
state_machine_id_in_shm = ++state_machines_count;
```
and in update_sm(), this cannot be static anymore if you do something like what I mentioned
```c++
state_machine_sm[state_machine_id_in_shm] = current_state;
```